### PR TITLE
Add the Gorilla model type

### PR DESCRIPTION
**Pull request by skejserjensen**

### CODE REVIEW

File: server/src/models/gorilla.rs [hunk 1]
```diff
@@ -264,3 +281,258 @@ impl<'a> Bits<'a> {
...
+        equal
+    }
+
+    fn equal_or_nan(v1: Value, v2: Value) -> bool {
+        v1 == v2 || (v1.is_nan() && v2.is_nan())
+    }
+
+    // Tests for BitReader.
+    #[test]
+    fn test_empty_bit_reader_error() {
```
Comment (CGodiksen) [inline server/src/models/gorilla.rs:482]: Maybe rename to "test_bit_reader_cannot_be_empty" or something similar for consistency with other `is_err()` tests.

File: server/src/models/gorilla.rs [hunk 2]
```diff
@@ -264,3 +281,258 @@ impl<'a> Bits<'a> {
         value as u32
     }
 }
+
+/// Append one or multiple bits to a `vec<u8>`.
+struct BitVecBuilder {
+    current_byte: u8,
```
Comment (CGodiksen) [inline server/src/models/gorilla.rs:287]: Maybe add doc comments to the struct fields for consistency.

**Review by CGodiksen — CHANGES_REQUESTED** [review]

File: server/src/models/gorilla.rs [hunk 3]
```diff
@@ -264,3 +281,258 @@ impl<'a> Bits<'a> {
...
+    fn append_a_zero_bit(&mut self) {
+        self.append_bits(0, 1)
+    }
+
+    /// Append a one bit to `bytes`.
+    fn append_a_one_bit(&mut self) {
+        self.append_bits(1, 1)
+    }
+
+    /// Append `number_of_bits` from `bits` to `bytes`.
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/models/gorilla.rs:311]: Changed to "the `BitVecBuilder`" as "bytes" is an implementation detail that I should probably not have mentioned in the documentation for the types interface.

File: server/src/models/gorilla.rs [hunk 4]
```diff
@@ -264,3 +281,258 @@ impl<'a> Bits<'a> {
...
+                self.remaining_bits = 8;
+            }
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.bytes.is_empty()
+    }
+
+    fn finnish(mut self) -> Vec<u8> {
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/models/gorilla.rs:340]: Fixed.

File: server/src/models/gorilla.rs [hunk 5]
```diff
@@ -264,3 +281,258 @@ impl<'a> Bits<'a> {
...
+
+            if self.remaining_bits == 0 {
+                self.bytes.push(self.current_byte);
+                self.current_byte = 0;
+                self.remaining_bits = 8;
+            }
+        }
+    }
+
+    fn is_empty(&self) -> bool {
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/models/gorilla.rs:336]: I agree that all functions/methods generally should have a doc comment. For some reason I overlooked `is_empty` and `finish`. Both methods now have doc comments.

File: server/src/models/gorilla.rs [hunk 6]
```diff
@@ -13,246 +13,263 @@
...
+        Self {
+            last_value: 0.0,
+            last_leading_zero_bits: u8::MAX,
+            last_trailing_zero_bits: 0,
+            compressed_values: BitVecBuilder::new(),
+        }
+    }
+
+    /// Compress `value` using XOR and a variable length binary encoding.
+    fn compress_value(&mut self, value: Value) {
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/models/gorilla.rs:61]: I purposely used different names for these methods as the operation they perform is different and we do not need the model types to use the same interface through a `trait`. If this becomes necessary, e.g., during development of the compressor, we will of course be forced to use the same generic names for these methods.

File: server/src/models/gorilla.rs [hunk 7]
```diff
@@ -13,246 +13,263 @@
...
+    decompress_values(start_time, end_time, sampling_interval, model, values);
 }
 
-/** Private Functions **/
-//TODO: can gorilla::decode be shared without allocating an array for min, max, etc?
-fn decode(
-    start_time: i64,
-    end_time: i64,
+/// Decompress values compressed using Gorilla's compression method for
+/// floating-point values and store them in a new `PrimitiveArray<Value>`.
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/models/gorilla.rs:186]: Changed to "... in a new Apache Arrow array."

File: server/src/models/gorilla.rs [hunk 8]
```diff
@@ -13,246 +13,263 @@
...
+        values.append_value(Value::from_bits(last_value)).unwrap();
     }
 }
 
-//Bits
-//The implementation of this Type is based on code published by Ilkka Rauta
-// under the MIT/Apache2 license. LINK: https://github.com/irauta/bitreader
-struct Bits<'a> {
+// BitReader is implemented based on code published by Ilkka Rauta under both
+// the MIT and Apache2 licenses. LINK: https://github.com/irauta/bitreader.
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/models/gorilla.rs:244]: Changed the formatting of the link and made the comment a doc comment so the attribution is visible in the documentation.

File: server/src/models/gorilla.rs [hunk 9]
```diff
@@ -13,246 +13,263 @@
...
-        values.append_value(f32::from_bits(last_value)).unwrap();
+        values.append_value(Value::from_bits(last_value)).unwrap();
     }
 }
 
-//Bits
-//The implementation of this Type is based on code published by Ilkka Rauta
-// under the MIT/Apache2 license. LINK: https://github.com/irauta/bitreader
-struct Bits<'a> {
+// BitReader is implemented based on code published by Ilkka Rauta under both
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/models/gorilla.rs:243]: Fixed.

File: server/src/models/gorilla.rs [hunk 10]
```diff
@@ -13,246 +13,263 @@
...
+// the MIT and Apache2 licenses. LINK: https://github.com/irauta/bitreader.
+/// Read one or multiple bits from a `[u8]`.
+struct BitReader<'a> {
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/models/gorilla.rs:246]: My intention was to move them to a separate module if they are ever needed by other modules. Until then, I suggest we keep we them in the same module as the `Gorilla` model type as it is the only type that is using them.

File: server/src/models/gorilla.rs [hunk 11]
```diff
@@ -13,246 +13,263 @@
...
     bytes: &'a [u8],
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/models/gorilla.rs:247]: Fixed.

File: server/src/models/gorilla.rs [hunk 12]
```diff
@@ -13,246 +13,263 @@
...
     current_bit: u64,
 }
 
-impl<'a> Bits<'a> {
-    pub fn new(bytes: &'a [u8]) -> Bits<'a> {
-        Self {
-            bytes,
-            current_bit: 0,
+impl<'a> BitReader<'a> {
+    fn try_new(bytes: &'a [u8]) -> Result<Self, String> {
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/models/gorilla.rs:252]: I agree and I have added it as a task to do when the `models` module is refactored.

File: server/src/models/gorilla.rs [hunk 13]
```diff
@@ -13,246 +13,263 @@
...
+        } else {
+            Ok(Self {
+                bytes,
+                current_bit: 0,
+            })
         }
     }
 
-    pub fn read_bit(&mut self) -> bool {
+    /// Read the next bit from `bytes`.
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/models/gorilla.rs:263]: Replaced "`bytes`" with "the `BitReader`" to not leak implementation details into the documentation for the types interface.

File: server/src/models/gorilla.rs [hunk 14]
```diff
@@ -13,246 +13,263 @@
...
+    fn new() -> Self {
+        Self {
+            last_value: 0.0,
+            last_leading_zero_bits: u8::MAX,
+            last_trailing_zero_bits: 0,
+            compressed_values: BitVecBuilder::new(),
+        }
+    }
+
+    /// Compress `value` using XOR and a variable length binary encoding.
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/models/gorilla.rs:60]: Fixed.

File: server/src/models/gorilla.rs (hunk 1 above)
Comment (skejserjensen) [reply server/src/models/gorilla.rs:482]: Fixed.

File: server/src/models/gorilla.rs (hunk 2 above)
Comment (skejserjensen) [reply server/src/models/gorilla.rs:287]: Fixed.

File: server/src/models/gorilla.rs [hunk 15]
```diff
@@ -264,3 +281,258 @@ impl<'a> Bits<'a> {
...
+    use proptest::{bool, collection, prop_assert, prop_assume, proptest};
+
+    // The largest byte, a random byte, and the smallest byte for testing.
+    const TEST_BYTES: &[u8] = &[255, 170, 0];
+    const TEST_BITS: &[bool] = &[
+        true, true, true, true, true, true, true, true, true, false, true, false, true, false,
+        true, false, false, false, false, false, false, false, false, false,
+    ];
+
+    // Tests for Gorilla.
```
[the remark this answers is not in the document]
Comment (skejserjensen) [reply server/src/models/gorilla.rs:369]: Added tests for `compress_value` to exercise each branch explicitly.

File: server/src/models/gorilla.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/src/models/gorilla.rs:62]: That makes sense.

File: server/src/models/gorilla.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (CGodiksen) [reply server/src/models/gorilla.rs:248]: I agree.

**Review by CGodiksen — APPROVED** [review]

**Pull request merged**